### DISCUSSION TITLE
17.0

### DISCUSSION THREAD

**Discussion by bbakerman** [graphql-java/graphql-java#2500] [General]
We are happy to announce v17.0 of graphql-java

https://github.com/graphql-java/graphql-java/milestone/31?closed=1

#  `@deprecated` supported on input fields

In line with coming graphql spec changes, the `@deprecated` annotation can now be placed on input fields and not just output fields as before.

https://github.com/graphql-java/graphql-java/pull/2186

# Dramatic performance improvements in `GraphQLSchema` building

The `GraphQLSchema` is a graph of types and as such has to be built according to strict rules.  The previous code would repeated visit and rebuilt the schema and this meant that large schemes could be slow to build.

This code has been extensively benchmarked and improved.  For example on a very large and complex schema, total build time went from 3.5 seconds to 1.2 seconds.  

Thank you to teams out there that used https://github.com/graphql-java/graphql-anonymizer to submit in real life schemas that we can test with.

# Dramatic performance improvements in `DataFetchingFieldSelectionSet`

The classes that underlie the `DataFetchingFieldSelectionSet` code have been been remodelled to be dramatically more performant.  Before interfaces could create an exponential explosion in objects created.  This has been fixed which in turn improves the performance of this code as scale.

# Dramatic performance improvements in large query validation

The graphql spec calls for overlapping fields in fragments to be validated in a certain way.  The previous code followed the specs psudeo-algorithm correctly but it leads to asymptotic performance problems.

This has been rewritten  via https://github.com/graphql-java/graphql-java/pull/2495 based on this article https://tech.xing.com/graphql-overlapping-fields-can-be-merged-fast-ea6e92e0a01 which dramatically improves how fast large queries can validated.

# Support of directives on fields and types during introspection 

#2221 this is a non spec feature which lets you expose directives through Introspection

# Reworked default value validation and improved Scalar coercion 

#2325 This is a relatively big change which fixed up how default values are processed. Before that default values were never validated and custom Scalars could cause invalid values printed in SDL or Introspection. This PR includes a new Scalar coercion function `valueToLiteral`  which is needed to implement this correct behaviour.

# Better Unicode support in documents

https://github.com/graphql-java/graphql-java/pull/2404 allows for better Unicode expression in queries and SDL documents

# GraphQLContext is now the approved context mechanism

https://github.com/graphql-java/graphql-java/pull/2368 allows for a well known API for context within a graphql request.  Previously any  opaque object could be used and framework could was not able to make any assumptions about adding their own context properties

The old mechanism has been deprecated and a singleton (buy mutable) context object is  now how context is passed around.

# Ignored characters are no longer help in memory by default

Characters in a graphql document that are not significant to the syntax, such as spaces, new lines and comms, used to be captured in memory against AST nodes.

This is no longer the case and it saves memory from parsed documents.  There is a switch to turn on the old behavior intended for tooling code.

https://github.com/graphql-java/graphql-java/pull/2394

# Exception handlers can now be asynchronous

graphql-java operates in a asynchronous world and hence the code that handles exceptions in data fetchers should itself allow asynchronous handling. 

https://github.com/graphql-java/graphql-java/pull/2371

# Breaking changes

https://github.com/graphql-java/graphql-java/pulls?q=is%3Apr+label%3A%22breaking+change%22++milestone%3A%2217.0%22

* The boolean values acceptable as variable input the Boolean scalar have been tightened to `true` and `false` - see https://github.com/graphql-java/graphql-java/pull/2419

* The non standard `@Fetch` directive support has been removed by default and deprecated.  This was done for performance reasons because very few people used it but it slowed down schema creation a lot

* The long deprecated  `BatchedExecutionStrategy` and `ExecutorServiceExecutionStrategy` have finally been removed

<hr /><em>This discussion was created from the release <a href='https://github.com/graphql-java/graphql-java/releases/tag/v17.0'>17.0</a>.</em>

**Comment by dfa1** [comment]
@bbakerman there is also a matching release for graphql-java-extended-scalars 17.0?

**Reply by dfa1** [reply]
by the way, we are in production with 17.0-beta1, without any issues... thank you guys! 

(now switching to 17.0 for next production candidate)

**Reply by bbakerman** [reply]
yes - extended-scalars and extended-validation have been released

**Reply by setchy** [reply]
Nice work! 🎉 

One other item worth added to the changelog under breaking is the [removal](https://github.com/graphql-java/graphql-java/commit/9658168f58b0f110ada025b8a836c5ca35c8e5e2#diff-67db841eb271342a26511c008af3ccca914b4e3fed2e8e28f0afadcda61a85d6) of non standard scalars in 17.0 (BigDecimal, BigInteger, Byte, Char, Long, Short).  

This was previously flagged in the [15.0 release notes](https://github.com/graphql-java/graphql-java/releases/tag/v15.0)

**Reply by bbakerman** [reply]
I have updated the release notes to say this

**Reply by setchy** [reply]
@bbakerman @andimarek - thanks for updating the release notes.

Looks like those 6 scalars never made their way over to the [graphql-java-extended-scalars project](https://github.com/graphql-java/graphql-java-extended-scalars/tree/master/src/main/java/graphql/scalars).  Is that the intention, and will those 6 scalars (BigDecimal, BigInteger, Byte, Char, Long and Short) be added in a 17.1.0 release of that project?

**Reply by bbakerman** [reply]
https://github.com/graphql-java/graphql-java-extended-scalars/blob/master/src/main/java/graphql/scalars/ExtendedScalars.java#L216

Umm there are there??  I just checked the 17.0 release jar and there are there

**Reply by setchy** [reply]
😅 my bad - didn't think to look in `JavaPrimitives`

To make up, I've updated the readme with [this PR](https://github.com/graphql-java/graphql-java-extended-scalars/pull/48)

**Comment by guykatz** [comment]
> We are happy to announce v17.0 of graphql-java

is this a LTS version?

**Reply by andimarek** [reply]
@guykatz we are not providing any specific LTS versions at the moment based on the community feedback and our limited resources. We may backport certain critically PRs, but in general we only support the latest version.

**Reply by andimarek** [reply]
See also https://www.graphql-java.com/blog/17-released-and-lts/

**Comment by phillip-kruger** [comment]
Congrats ! Trying it out now :)

**Comment by phillip-kruger** [comment]
@bbakerman @andimarek thanks for your work on this release ! 
Initial tests in SmallRye GraphQL seems ok, but I am seeing that input String scalars are now being created in the Schema with a empty default

Before:
```
foo(bar: String): String
```

Now:
```
foo(bar: String = ): String
```

Any pointers ? Can I switch back to the old way ? 

Thanks in advance :)

**Reply by andimarek** [reply]
This looks wrong. In fact the printed SDL is not valid. Please try to reproduce it in a pure Graphql Java test, if possible.

In general we did a lot work around default values and Scalars which included breaking changes (see release note for details). Maybe this is related to it.

**Reply by bbakerman** [reply]
>  input String scalars are now being created in the Schema with a empty default

Can you please give us more details on the code creating these.  We suspect it is not doing it the correct way for 17.0

Links to github code ec... would be useful

A test showing this happening as a PR would be unreal

**Reply by phillip-kruger** [reply]
Thanks for the reply. I am playing around today to see if I can pin point it

**Reply by phillip-kruger** [reply]
Ok, I managed to find the issue, and I can work around it.

When creating the argument, we set the `defaultValue` (that I see is now deprecated, I assume needs to be replaced with `defaultValueProgrammatic`). This value can sometimes be null.

So when doing:

```
GraphQLArgument.Builder argumentBuilder = GraphQLArgument.newArgument()
                .name("bar")
                .defaultValue(null);
```

will create 

```
type Query {
  foo(bar: String = ): String
}
```

Changing to `defaultValueProgrammatic` :

```
GraphQLArgument.Builder argumentBuilder = GraphQLArgument.newArgument()
                .name("bar")
                .defaultValueProgrammatic(null);
```
gives

```
type Query {
  foo(bar: String = null): String
}
```

So to get back to what it was:

```
GraphQLArgument.Builder argumentBuilder = GraphQLArgument.newArgument()
                .name("bar");
        
        if(defaultValue!=null){
            argumentBuilder.defaultValueProgrammatic(defaultValue);
        }  
```
So that is ok for me, I can code around it.

Thanks :)

**Comment by davenice** [comment]
Hey,

I'm a bit late to the party on this but moving up versions in our graphql-java implementation.

We have a fairly complex schema which is programmatically generated. In our unit tests we have small snippets of the schema to test again.

In some of these snippets, we have ObjectA in our schema which refers to ObjectB. ObjectB is not in our schema in the unit test. At graphql-java 16, we often managed to get away with this. Moving up to graphql-java 17, we've hit a few test failures like:
```
testEmptyAggregateRecordIsNull(com.xxx.graphql.MyTest)  Time elapsed: 8.314 sec  <<< ERROR!
graphql.AssertException: type ObjectB not found in schema
	at graphql.Assert.assertNotNull(Assert.java:17)
	at graphql.schema.GraphQLTypeResolvingVisitor.handleTypeReference(GraphQLTypeResolvingVisitor.java:49)
	at graphql.schema.GraphQLTypeResolvingVisitor.visitGraphQLTypeReference(GraphQLTypeResolvingVisitor.java:44)
	at graphql.schema.GraphQLTypeReference.accept(GraphQLTypeReference.java:62)
	at graphql.schema.SchemaTraverser$TraverserDelegateVisitor.enter(SchemaTraverser.java:87)
	at graphql.util.Traverser.traverse(Traverser.java:144)
	at graphql.schema.SchemaTraverser.doTraverse(SchemaTraverser.java:74)
	at graphql.schema.SchemaTraverser.depthFirst(SchemaTraverser.java:64)
	at graphql.schema.SchemaTraverser.depthFirst(SchemaTraverser.java:57)
	at graphql.schema.impl.SchemaUtil.replaceTypeReferences(SchemaUtil.java:97)
	at graphql.schema.GraphQLSchema$Builder.buildImpl(GraphQLSchema.java:821)
	at graphql.schema.GraphQLSchema$Builder.build(GraphQLSchema.java:787)
	at com.xxx.graphql.MyTest.createSchema(MyTest.java:84)
```

Is this a known side effect of the performance improvements in graphql-java 17? Is there a flag I can set to encourage graphql-java to be more tolerant of the schema in unit tests?

Thanks in advance!

**Reply by dondonz** [reply]
Hi @davenice could you please provide a minimal example to reproduce the error? That would help me debug the issue. Thanks!

**Reply by davenice** [reply]
Thanks @dondonz - in the meantime I've just put more of my schema in the unit tests. Because it's almost 2 years out in the field I imagine if people were generally getting his by this you'd know by now! I just wanted to check whether there was something that came to mind - or if there was a parameter to control strictness!

I don't think it would be a good of your time to go debugging. :-)

**Reply by dondonz** [reply]
Bugs can definitely last longer than 2 years 😄 so always feel free to report issues!
